### PR TITLE
Simple RPC response caching

### PR DESCRIPTION
**Pull request by nathanwoctopusdeploy**
# Background

This PR allows a Halibut client to cache responses from the Service for selected RPC calls.

The interface used by the client can have an attribute applied that instructs Halibut to cache the response from the service for the specified duration in seconds.

```
public interface IMyService
{
    [CacheResponse(600)]
    Guid CachableCall();
}
```

The responses are cached in a `System.Runtime.Caching.MemoryCache`

By default, only non-error responses are cached.

## Cache Validity

The cached items are expired from the cache after the absolute cache duration specified on the `CacheResponseAtttribute`. No other cache invalidation is performed.

The implementor must ensure that caching responses from a service call make sense and that any possible validity issues arising from caching are handled. Halibut has no logic to invalidate a cache other than at the absolute expiration of a cached item.

## Caching Error Responses

It is possible to cache an error response by overriding the cache behaviour on an error-by-error basis. This allows for scenarios such as caching the error response when a `NoMatchingServiceOrMethodHalibutClientException` exception is thrown.

```
IHalibutRuntime halibut = ....;
halibut.OverrideErrorResponseMessageCaching = (response) =>
{
    if (response.Error.Message.Contains("CACHE ME"))
    {
        return true;
    }

    return false;
};
```

# How to review this PR

Quality :heavy_check_mark:

# Pre-requisites

- [ ] I have read [How we use GitHub Issues](https://github.com/OctopusDeploy/Issues/blob/master/docs/CONTRIBUTING.internal.md) for help deciding when and where it's appropriate to make an issue.
- [ ] I have considered informing or consulting the right people, according to the [ownership map](https://whimsical.com/ownership-map-NzbiD4HJyvhC9jNJNfS6TG).
- [ ] I have considered appropriate testing for my change.

### CODE REVIEW

File: source/Halibut.Tests/ResponseMessageCacheFixture.cs [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+
+        [Test]
+        [TestCaseSource(nameof(ServiceConnectionTypeAndVersion))]
+        public async Task ForAServiceThatSupportsCaching_ClientShouldBeAbleToForceSpecificErrorResponsesToBeCached(ServiceConnectionType serviceConnectionType, string halibutServiceVersion)
+        {
+            using var clientAndService = await CreateClientAndService(serviceConnectionType, halibutServiceVersion);
+            {
+                clientAndService.Octopus.OverrideErrorResponseMessageCaching = (response) =>
+                {
+                    if (response.Error.Message.Contains("CACHE ME"))
```
Comment (LukeButters) [inline source/Halibut.Tests/ResponseMessageCacheFixture.cs:177]: nit: just return `response.Error.Message.Contains("CACHE ME")`

File: source/Halibut.Tests/ResponseMessageCacheFixture.cs [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+
+                    return false;
+                };
+
+                var client = clientAndService.CreateClient<ICachingService>();
+
+                Exception exception1 = null;
+                Exception exception2 = null;
+
+                TryCatch(() => client.CachableCallThatThrowsAnException($"UNCACHED"), e => exception1 = e);
```
Comment (LukeButters) [inline source/Halibut.Tests/ResponseMessageCacheFixture.cs:190]: nit: I wonder if  a should/could throw assertion could be used here. I think that would mean you end up with:
```
var exception1 = Assert.shouldThrow(() => client.CachableCallThatThrowsAnException($"UNCACHED"))
```

Removing the need for the TryCatch method and removes the need for the null check.

File: source/Halibut.TestUtils.CompatBinary.Base/ICachingService.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        // It is applied on the client interface in the Halibut.Tests project
+        Guid CachableCall(Guid input);
+
+        // The CacheResponse Attribute cannot be applied / isn't need for the Service interface
+        // It is applied on the client interface in the Halibut.Tests project
+        Guid AnotherCachableCall();
+
+        // The CacheResponse Attribute cannot be applied / isn't need for the Service interface
+        // It is applied on the client interface in the Halibut.Tests project
+        Guid CachableCallThatThrowsAnException(string exceptionMessage);
```
Comment (LukeButters) [inline source/Halibut.TestUtils.CompatBinary.Base/ICachingService.cs:23]: nit: Consider renaming this to `CachableCallThatThrowsAnExceptionWithARandomExceptionMessage`.

also rename `exceptionMessage` to `exceptionMessagePrefix`

File: source/Halibut.Tests/ResponseMessageCacheFixture.cs [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+
+                exception3.Should().NotBeNull();
+                exception4.Should().NotBeNull();
+                exception3!.Message.Should().Be(exception4!.Message);
+            }
+        }
+
+        [Test]
+        [TestCaseSource(nameof(ServiceConnectionTypeAndVersion))]
+        public async Task ForAServiceThatSupportsCaching_ErrorResponsesShouldNotBeCached(ServiceConnectionType serviceConnectionType, string halibutServiceVersion)
```
Comment (LukeButters) [inline source/Halibut.Tests/ResponseMessageCacheFixture.cs:212]: nit: `ForAServiceThatSupportsCaching_ErrorResponsesShouldNotBeCached` -> `ForAServiceThatSupportsCaching_ErrorResponsesShouldNotBeCachedByDefault`

**Comment by nathanwoctopusdeploy** [comment]
[sc-46261]

File: source/Halibut/Transport/Caching/CachingKeyGenerator.cs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+        public string GetCacheKey(MethodInfo methodInfo, object[] args, string prefix)
+        {
+            var methodArguments = args?.Any() == true
+                ? args.Select(ParameterCacheKeys.GenerateCacheKey)
+                : new[] { "0" };
+            return GenerateCacheKey(methodInfo, prefix, methodArguments);
+        }
+
+        public string GetCacheKeyPrefix(MethodInfo methodInfo, string prefix)
```
Comment (LukeButters) [inline source/Halibut/Transport/Caching/CachingKeyGenerator.cs:21]: nit: make this internal

**Review by LukeButters — APPROVED** [review]

File: source/Halibut.Tests/CancellationViaClientProxyFixture.cs [hunk 6]
```diff
@@ -55,18 +55,29 @@ public void CannotHaveServiceWithHalibutProxyRequestOptions()
         [TestCase(ServiceConnectionType.Polling)]
         public async Task CanTalkToOldServicesWhichDontKnowAboutHalibutProxyRequestOptions(ServiceConnectionType serviceConnectionType)
         {
-            using (var clientAndService = await ClientAndPreviousVersionServiceBuilder.WithService(serviceConnectionType).WithServiceVersion("5.0.429").Build())
+            using (var clientAndService = await ClientAndPreviousServiceVersionBuilder.WithService(serviceConnectionType).WithServiceVersion("5.0.429").Build())
```
Comment (acodrington) [inline source/Halibut.Tests/CancellationViaClientProxyFixture.cs:58]: Can use the new `PreviousServiceVersions` constant here

**Review by acodrington — COMMENTED** [review]
:shipit:

**Review by acodrington — APPROVED** [review]

**Comment by nathanwoctopusdeploy** [comment]
Fixes: https://github.com/OctopusDeploy/Issues/issues/8229

**Pull request merged**